commit 463b1a107d0357787ccc0e032f822d8eb4baa75b
Author: Rob Johnston <rob.johnston@joyent.com>
Date:   2019-04-29T21:21:59+00:00 (5 months ago)
    
    OS-7776 Hitting enter before timeout causes loader to always boot from stick

diff --git a/usr/src/boot/sys/boot/forth/menu.4th b/usr/src/boot/sys/boot/forth/menu.4th
index 814445fac8..194dfec9a8 100644
--- a/usr/src/boot/sys/boot/forth/menu.4th
+++ b/usr/src/boot/sys/boot/forth/menu.4th
@@ -739,6 +739,21 @@ also menu-infrastructure definitions
 	then
 ;
 
+: do_ipxe ( -- bool)
+	\ getenv? leaves -1 on stack if the env var exists.  Thus if both
+	\ headnode and ipxe exist then the sum of what will be left on the
+	\ stack should be -2.
+	s" headnode" getenv? s" ipxe" getenv? + -2 = if
+		s" ipxe" getenv s" true" compare 0= if
+			true
+		else
+			false
+		then
+	else
+		false
+	then
+;
+
 \ Takes a single integer on the stack and updates the timeout display. The
 \ integer must be between 0 and 9 (we will only update a single digit in the
 \ source message).
@@ -749,11 +764,8 @@ also menu-infrastructure definitions
 	dup 9 > if drop 9 then
 	dup 0 < if drop 0 then
 
-	\ getenv? leaves -1 on stack if the env var exists.  Thus if both
-	\ headnode and ipxe exist then the sum of what will be left on the
-	\ stack should be -2.
-	s" headnode" getenv? s" ipxe" getenv? + -2 = if
-		s" ipxe" getenv s" true" compare 0= if
+	s" headnode" getenv? if
+		do_ipxe if
 			s" Autoboot in N seconds from PXE. [Space] to pause" ( n -- n c-addr/u )
 		else
 			s" Autoboot in N seconds from the USB Key. [Space] to pause" ( n -- n c-addr/u )
@@ -1208,8 +1220,16 @@ also menu-namespace
 		\ Ctrl-Enter/Ctrl-J (10)
 		dup over 13 = swap 10 = or if
 			drop ( no longer needed )
-			s" boot" evaluate
-			exit ( pedantic; never reached )
+			do_ipxe if
+				s" efi-version" getenv? if
+					s" ipxe_chainload" evaluate
+				else
+					s" ipxe_boot" evaluate
+				then
+			else
+				s" boot" evaluate
+				exit ( pedantic; never reached )
+			then
 		then
 
 		dup menureboot @ = if 0 reboot then
